File: temp_zip_linux/lib/protocol/sos_envelope.dart
import 'dart:convert';
import 'dart:collection';

import '../sos_kernel.dart';

class SosEnvelope {
  final int version;
  final String type;
  final String sender;
  final int timestamp;
  final Map<String, dynamic> payload;
  final String signature;

  SosEnvelope({
    required this.version,
    required this.type,
    required this.sender,
    required this.timestamp,
    required this.payload,
    required this.signature,
  });

  Map<String, dynamic> toJson() => {
        'v': version,
        'type': type,
        'sender': sender,
        'timestamp': timestamp,
        'payload': payload,
        'signature': signature,
      };

  String toJsonString() => jsonEncode(toJson());

  String? get target {
    final value = payload['target'];
    return value == null ? null : value.toString();
  }

  bool get isBroadcast {
    final t = target;
    return t == null || t.isEmpty || t == '*';
  }

  static SosEnvelope fromJson(Map<String, dynamic> json) {
    return SosEnvelope(
      version: json['v'] as int? ?? 1,
      type: json['type'] as String,
      sender: json['sender'] as String,
      timestamp: json['timestamp'] as int,
      payload: Map<String, dynamic>.from(json['payload'] as Map),
      signature: json['signature'] as String,
    );
  }

  static SosEnvelope fromJsonString(String raw) {
    final data = jsonDecode(raw) as Map<String, dynamic>;
    return SosEnvelope.fromJson(data);
  }

  static Future<SosEnvelope> sign({
    required SosCore core,
    required String type,
    required Map<String, dynamic> payload,
  }) async {
    final timestamp = DateTime.now().millisecondsSinceEpoch;
    final sender = core.publicKey;
    final body = _canonicalBody(
      version: 1,
      type: type,
      sender: sender,
      timestamp: timestamp,
      payload: payload,
    );
    final signature = await core.sign(body);
    return SosEnvelope(
      version: 1,
      type: type,
      sender: sender,
      timestamp: timestamp,
      payload: payload,
      signature: signature,
    );
  }

  Future<bool> verifyWith(SosCore core) async {
    final body = _canonicalBody(
      version: version,
      type: type,
      sender: sender,
      timestamp: timestamp,
      payload: payload,
    );
    return core.verifySignature(
      message: body,
      signatureB64: signature,
      publicKeyB64: sender,
    );
  }

  static String _canonicalBody({
    required int version,
    required String type,
    required String sender,
    required int timestamp,
    required Map<String, dynamic> payload,
  }) {
    final body = {
      'v': version,
      'type': type,
      'sender': sender,
      'timestamp': timestamp,
      'payload': payload,
    };
    return jsonEncode(_normalize(body));
  }

  String canonicalBody() => _canonicalBody(
        version: version,
        type: type,
        sender: sender,
        timestamp: timestamp,
        payload: payload,
      );

  static Object _normalize(Object value) {
    if (value is Map) {
      final sorted = SplayTreeMap<String, dynamic>();
      value.forEach((key, v) {
        sorted[key.toString()] = _normalize(v);
      });
      return sorted;
    }
    if (value is List) {
      return value.map((item) => _normalize(item)).toList();
    }
    return value;
  }
}

File: java/core/src/main/java/com/resgatesos/core/SosEnvelope.java
package com.resgatesos.core;

import com.google.gson.Gson;
import com.google.gson.JsonObject;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public class SosEnvelope {
    public int v;
    public String type;
    public String sender;
    public long timestamp;
    public Map<String, Object> payload;
    public String signature;

    public SosEnvelope() {
    }

    public SosEnvelope(int v, String type, String sender, long timestamp,
            Map<String, Object> payload, String signature) {
        this.v = v;
        this.type = type;
        this.sender = sender;
        this.timestamp = timestamp;
        this.payload = payload;
        this.signature = signature;
    }

    public static SosEnvelope sign(CryptoManager crypto, String type, Map<String, Object> payload) throws Exception {
        long ts = System.currentTimeMillis();
        String sender = crypto.getPublicKeyBase64();
        String body = canonicalBody(1, type, sender, ts, payload);
        String sig = crypto.sign(body);
        return new SosEnvelope(1, type, sender, ts, payload, sig);
    }

    public boolean verify(CryptoManager crypto) {
        String body = canonicalBody(v, type, sender, timestamp, payload);
        return crypto.verify(body, signature, sender);
    }

    public String toJsonString() {
        return new Gson().toJson(toJson());
    }

    public Map<String, Object> toJson() {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("v", v);
        data.put("type", type);
        data.put("sender", sender);
        data.put("timestamp", timestamp);
        data.put("payload", payload);
        data.put("signature", signature);
        return data;
    }

    public String canonicalBody() {
        return canonicalBody(v, type, sender, timestamp, payload);
    }

    public static SosEnvelope fromJsonString(String raw) {
        JsonObject obj = new Gson().fromJson(raw, JsonObject.class);
        SosEnvelope envelope = new SosEnvelope();
        envelope.v = obj.get("v").getAsInt();
        envelope.type = obj.get("type").getAsString();
        envelope.sender = obj.get("sender").getAsString();
        envelope.timestamp = obj.get("timestamp").getAsLong();
        envelope.signature = obj.get("signature").getAsString();
        envelope.payload = new Gson().fromJson(obj.get("payload"),
                new com.google.gson.reflect.TypeToken<Map<String, Object>>() {
                }.getType());
        return envelope;
    }

    public String getTarget() {
        Object target = payload != null ? payload.get("target") : null;
        return target == null ? null : target.toString();
    }

    public boolean isBroadcast() {
        String t = getTarget();
        return t == null || t.isEmpty() || "*".equals(t);
    }

    private static String canonicalBody(int v, String type, String sender, long timestamp,
            Map<String, Object> payload) {
        Map<String, Object> body = new HashMap<>();
        body.put("v", v);
        body.put("type", type);
        body.put("sender", sender);
        body.put("timestamp", timestamp);
        body.put("payload", payload);
        return new Gson().toJson(normalize(body));
    }

    private static Object normalize(Object value) {
        if (value instanceof Map) {
            TreeMap<String, Object> sorted = new TreeMap<>();
            ((Map<?, ?>) value).forEach((k, v) -> sorted.put(k.toString(), normalize(v)));
            return sorted;
        }
        if (value instanceof List) {
            return ((List<?>) value).stream().map(SosEnvelope::normalize).toList();
        }
        return value;
    }
}
